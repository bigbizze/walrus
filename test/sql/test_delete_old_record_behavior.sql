/*

*/
select 1 from pg_create_logical_replication_slot('realtime', 'wal2json', false);

create table public.notes(
    pk1 int primary key,
    body text
);

insert into realtime.subscription(subscription_id, entity, claims)
select
    seed_uuid(1),
    'public.notes',
    jsonb_build_object(
        'role', 'authenticated',
        'email', 'example@example.com',
        'sub', seed_uuid(1)::text
    );

-- Option 1:
-- Replica Identity Full: false
-- Row Level Security   : false

-- Expect:
-- old_record contains only primary key info because only pkey info available in WAL
alter table public.notes replica identity default;
alter table public.notes disable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
delete from public.notes where pk1=1;

select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;

-- Option 2:
-- Replica Identity Full: false
-- Row Level Security   : true

-- Expect:
-- old_record contains only primary key info because only pkey info available in WAL
alter table public.notes replica identity default;
alter table public.notes enable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
delete from public.notes where pk1=1;

select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;


-- Option 3:
-- Replica Identity Full: true
-- Row Level Security   : false

-- Expect:
-- old_record contains all columns becaues they're all available and there is no RLS so info is public
alter table public.notes replica identity full;
alter table public.notes disable row level security;

insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
delete from public.notes where pk1=1;

select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;

-- Option 4:
-- Replica Identity Full: true
-- Row Level Security   : true

-- Expect:
-- old_record contains only primary key info because we can not enforce RLS on deletes and some columns might be private
alter table public.notes replica identity full;
alter table public.notes enable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
delete from public.notes where pk1=1;

select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;

select pg_drop_replication_slot('realtime');
drop table public.notes;
truncate table realtime.subscription;
