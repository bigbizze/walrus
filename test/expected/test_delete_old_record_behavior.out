/*

*/
select 1 from pg_create_logical_replication_slot('realtime', 'wal2json', false);
 ?column? 
----------
        1
(1 row)

create table public.notes(
    pk1 int primary key,
    body text
);
insert into realtime.subscription(subscription_id, entity, claims)
select
    seed_uuid(1),
    'public.notes',
    jsonb_build_object(
        'role', 'authenticated',
        'email', 'example@example.com',
        'sub', seed_uuid(1)::text
    );
-- Option 1:
-- Replica Identity Full: false
-- Row Level Security   : false
-- Expect:
-- old_record contains only primary key info because only pkey info available in WAL
alter table public.notes replica identity default;
alter table public.notes disable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
 clear_wal 
-----------
 
(1 row)

delete from public.notes where pk1=1;
select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;
                      rec                       | is_rls_enabled |            subscription_ids            | errors 
------------------------------------------------+----------------+----------------------------------------+--------
 {                                             +| f              | {f4539ebe-c779-5788-bbc1-2421ffaa8954} | {}
     "type": "DELETE",                         +|                |                                        | 
     "table": "notes",                         +|                |                                        | 
     "schema": "public",                       +|                |                                        | 
     "columns": [                              +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "pk1",                    +|                |                                        | 
             "type": "int4"                    +|                |                                        | 
         },                                    +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "body",                   +|                |                                        | 
             "type": "text"                    +|                |                                        | 
         }                                     +|                |                                        | 
     ],                                        +|                |                                        | 
     "old_record": {                           +|                |                                        | 
         "pk1": 1                              +|                |                                        | 
     },                                        +|                |                                        | 
     "commit_timestamp": "2000-01-01T00:01:01Z"+|                |                                        | 
 }                                              |                |                                        | 
(1 row)

-- Option 2:
-- Replica Identity Full: false
-- Row Level Security   : true
-- Expect:
-- old_record contains only primary key info because only pkey info available in WAL
alter table public.notes replica identity default;
alter table public.notes enable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
 clear_wal 
-----------
 
(1 row)

delete from public.notes where pk1=1;
select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;
                      rec                       | is_rls_enabled |            subscription_ids            | errors 
------------------------------------------------+----------------+----------------------------------------+--------
 {                                             +| t              | {f4539ebe-c779-5788-bbc1-2421ffaa8954} | {}
     "type": "DELETE",                         +|                |                                        | 
     "table": "notes",                         +|                |                                        | 
     "schema": "public",                       +|                |                                        | 
     "columns": [                              +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "pk1",                    +|                |                                        | 
             "type": "int4"                    +|                |                                        | 
         },                                    +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "body",                   +|                |                                        | 
             "type": "text"                    +|                |                                        | 
         }                                     +|                |                                        | 
     ],                                        +|                |                                        | 
     "old_record": {                           +|                |                                        | 
         "pk1": 1                              +|                |                                        | 
     },                                        +|                |                                        | 
     "commit_timestamp": "2000-01-01T00:01:01Z"+|                |                                        | 
 }                                              |                |                                        | 
(1 row)

-- Option 3:
-- Replica Identity Full: true
-- Row Level Security   : false
-- Expect:
-- old_record contains all columns becaues they're all available and there is no RLS so info is public
alter table public.notes replica identity full;
alter table public.notes disable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
 clear_wal 
-----------
 
(1 row)

delete from public.notes where pk1=1;
select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;
                      rec                       | is_rls_enabled |            subscription_ids            | errors 
------------------------------------------------+----------------+----------------------------------------+--------
 {                                             +| f              | {f4539ebe-c779-5788-bbc1-2421ffaa8954} | {}
     "type": "DELETE",                         +|                |                                        | 
     "table": "notes",                         +|                |                                        | 
     "schema": "public",                       +|                |                                        | 
     "columns": [                              +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "pk1",                    +|                |                                        | 
             "type": "int4"                    +|                |                                        | 
         },                                    +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "body",                   +|                |                                        | 
             "type": "text"                    +|                |                                        | 
         }                                     +|                |                                        | 
     ],                                        +|                |                                        | 
     "old_record": {                           +|                |                                        | 
         "pk1": 1,                             +|                |                                        | 
         "body": "take out trash"              +|                |                                        | 
     },                                        +|                |                                        | 
     "commit_timestamp": "2000-01-01T00:01:01Z"+|                |                                        | 
 }                                              |                |                                        | 
(1 row)

-- Option 4:
-- Replica Identity Full: true
-- Row Level Security   : true
-- Expect:
-- old_record contains only primary key info because we can not enforce RLS on deletes and some columns might be private
alter table public.notes replica identity full;
alter table public.notes enable row level security;
insert into public.notes(pk1, body) values (1, 'take out trash');
select clear_wal();
 clear_wal 
-----------
 
(1 row)

delete from public.notes where pk1=1;
select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;
                      rec                       | is_rls_enabled |            subscription_ids            | errors 
------------------------------------------------+----------------+----------------------------------------+--------
 {                                             +| t              | {f4539ebe-c779-5788-bbc1-2421ffaa8954} | {}
     "type": "DELETE",                         +|                |                                        | 
     "table": "notes",                         +|                |                                        | 
     "schema": "public",                       +|                |                                        | 
     "columns": [                              +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "pk1",                    +|                |                                        | 
             "type": "int4"                    +|                |                                        | 
         },                                    +|                |                                        | 
         {                                     +|                |                                        | 
             "name": "body",                   +|                |                                        | 
             "type": "text"                    +|                |                                        | 
         }                                     +|                |                                        | 
     ],                                        +|                |                                        | 
     "old_record": {                           +|                |                                        | 
         "pk1": 1                              +|                |                                        | 
     },                                        +|                |                                        | 
     "commit_timestamp": "2000-01-01T00:01:01Z"+|                |                                        | 
 }                                              |                |                                        | 
(1 row)

select pg_drop_replication_slot('realtime');
 pg_drop_replication_slot 
--------------------------
 
(1 row)

drop table public.notes;
truncate table realtime.subscription;
